includes/widgets/class-wc-widget-layered-nav.php

```diff
@@ -409,9 +410,36 @@ protected function get_filtered_term_product_counts( $term_ids, $taxonomy, $quer
 			$query['where'] .= ' AND ' . $search;
 		}
 
-		$query['group_by'] = 'GROUP BY terms.term_id';
-		$query             = apply_filters( 'woocommerce_get_filtered_term_product_counts_query', $query );
-		$query_sql         = implode( ' ', $query );
+		$query          = apply_filters( 'woocommerce_get_filtered_term_product_counts_query', $query );
+		$main_query_sql = implode( ' ', $query );
+
+		// Generate the second part of the query.
+		// This one will return products having "Any..." as the value of the attribute.
+
+		$query_sql_for_attributes_with_any_value = "
+			SELECT {$wpdb->posts}.post_parent AS product_id, {$wpdb->term_relationships}.term_taxonomy_id as term_count_id FROM {$wpdb->postmeta}
+			JOIN {$wpdb->posts} ON {$wpdb->posts}.ID = {$wpdb->postmeta}.post_id
+			JOIN {$wpdb->term_relationships} ON {$wpdb->term_relationships}.object_id = {$wpdb->posts}.post_parent
+			WHERE {$wpdb->postmeta}.meta_key = 'attribute_$taxonomy'
+			AND {$wpdb->postmeta}.meta_value = ''
+			AND {$wpdb->posts}.post_type = 'product_variation'
+			AND {$wpdb->posts}.post_status = 'publish'
+			AND {$wpdb->term_relationships}.term_taxonomy_id in $term_ids_sql
+			AND NOT EXISTS (
+		        SELECT ID FROM {$wpdb->posts} AS parent
+		        WHERE parent.ID = {$wpdb->posts}.post_parent AND parent.post_status NOT IN ('publish')
+		    )
+			{$main_tax_query_sql['where']}";
+
+		// Generate the final query as the union+count of both.
+		$query_sql = "
+			SELECT COUNT(DISTINCT(product_id)) AS term_count, term_count_id FROM (
+				{$main_query_sql}
+				UNION ALL
```

Comment:
Instead of unionizing this, can we fetch the result in two separate queries and then combine. Basically group by over union is forcing MySQL to do a filesort with both the results, also this can have a large insert locking footprint.

I think we should favor performance here instead of atomicity.

Reply:
It's worth mentioning that we're caching this in a transient so if it's a little bit slower it isn't necessarily a significant problem so long as the locking footprint isn't against the posts table. @Konamiman and @vedanshujain what do you think of using this instead of a union or two queries:

```php
// Generate query.
$query           = array();
$query['select'] = "SELECT COUNT( DISTINCT {$wpdb->posts}.ID ) as term_count, terms.term_id as term_count_id";
$query['from']   = "FROM {$wpdb->posts}";
$query['join']   = "
	INNER JOIN {$wpdb->term_relationships} AS tr ON {$wpdb->posts}.ID = tr.object_id OR {$wpdb->posts}.post_parent = tr.object_id
	INNER JOIN {$wpdb->term_taxonomy} AS term_taxonomy ON tr.term_taxonomy_id = term_taxonomy.term_taxonomy_id
	LEFT JOIN (
		SELECT term_taxonomy_sub.taxonomy, tr_sub.object_id, term_taxonomy_sub.term_id FROM {$wpdb->term_relationships} AS tr_sub
		INNER JOIN {$wpdb->term_taxonomy} AS term_taxonomy_sub ON tr_sub.term_taxonomy_id = term_taxonomy_sub.term_taxonomy_id
	) AS tr_variation ON {$wpdb->posts}.ID = tr_variation.object_id AND term_taxonomy.taxonomy = tr_variation.taxonomy
	INNER JOIN {$wpdb->terms} AS terms
		ON tr_variation.term_id IS NULL AND term_taxonomy.term_id = terms.term_id
		OR tr_variation.term_id = terms.term_id
	{$main_tax_query_sql['join']} {$meta_query_sql['join']}"; // Not an omission, really no more JOINs required.
```

Reply:
We discussed this offline, and the query suggested by @ObliviousHarmony is currently non-optimized, so going ahead with a different query added in https://github.com/woocommerce/woocommerce/pull/27508/commits/07b62dabbde388d17d9e7e772de4e8b29b10d410